public/app/containers/Explore/Table.tsx

```diff
@@ -1,22 +1,58 @@
 import React, { PureComponent } from 'react';
-// import TableModel from 'app/core/table_model';
+import TableModel from 'app/core/table_model';
 
-const EMPTY_TABLE = {
-  columns: [],
-  rows: [],
+const EMPTY_TABLE = new TableModel();
+
+interface TableProps {
+  className?: string;
+  data: TableModel;
+  onClickCell?: (columnKey: string, rowValue: string) => void;
+}
+
+interface SFCCellProps {
+  columnIndex: number;
+  onClickCell?: (columnKey: string, rowValue: string, columnIndex: number, rowIndex: number, table: TableModel) => void;
+  rowIndex: number;
+  table: TableModel;
+  value: string;
+}
+
+const Cell: React.SFC<SFCCellProps> = props => {
+  const { columnIndex, rowIndex, table, value, onClickCell } = props;
+  const column = table.columns[columnIndex];
+  if (column && column.filterable && onClickCell) {
+    const onClick = event => {
+      event.preventDefault();
+      onClickCell(column.text, value, columnIndex, rowIndex, table);
+    };
+    return (
+      <td>
+        <a className="link" onClick={onClick}>
+          {value}
+        </a>
+      </td>
+    );
+  }
+  return <td>{value}</td>;
 };
 
-export default class Table extends PureComponent<any, any> {
+export default class Table extends PureComponent<TableProps, {}> {
```

Comment:
Sorry for my cavil, I'm new to React and want to figure out how to write good code. In this case, does PureComponent really provide any benefits? `TableProps` contains `data` field which is an object, so shallow comparison doesn't make sense for it I think.
https://reactjs.org/docs/react-api.html#reactpurecomponent
>Note
>React.PureComponent’s shouldComponentUpdate() only shallowly compares the objects. If these contain complex data structures, it may produce false-negatives for deeper differences. Only extend PureComponent when you expect to have simple props and state, or use forceUpdate() when you know deep data structures have changed. Or, consider using immutable objects to facilitate fast comparisons of nested data.

Reply:
Yes, it still makes sense because `data` is created new from each query response, so shallow object identity comparison will result in `false`, triggering an update. An optimisation could check if the data in the responses are the same, and then short-circuit, but I think that's an over-optimisation.

Reply:
Yes, it makes sense, thanks for the clarification!